### DISCUSSION TITLE
Unexpected behaviour of `FileInterface::getContents` method in `reactphp/filesystem`

### DISCUSSION THREAD

**Discussion by odmyikona** [reactphp/reactphp#509] [Q&A]
Hi there,

Thanks once again for this wonderful software.

Is it the expected behaviour of `FileInterface::getContents` method not to reflect changes if it is called once or am I missing something?

Under this series of events the result of the final step is different from what's expected:

- Running a loop
- Get file content with `FileInterface::getContents`
- Append to file with `FileInterface::putContents` (using await)
- Get file contents again with `FileInterface::getContents`, but the content do not reflect the changes made in previous step

Note: `file_get_contents` works.

I am running "Framework-X" in a docker container as such:

```Dockerfile
FROM php:8.1-alpine3.16

RUN apk --no-cache add ${PHPIZE_DEPS} libev \
    && pecl install ev \
    && docker-php-ext-enable ev \
    && docker-php-ext-install sockets \
    && docker-php-ext-install pcntl \
    && echo "memory_limit = -1" >> "${PHP_INI_DIR}/conf.d/acme.ini"

[...]
 ```
Thanks!

**Comment by WyriHaximus** [comment]
Hey @odmyikona,

Glad to hear your enjoy using our project :D

Going to assume you're using the `0.2` branch. The `0.2` branch is doing a `stat` before a `read`, currently working on a PR that will remove that and test a few situations that might occur while reading from a file. The issue could be that that `stat` is cached and as such still thinks the file is the old size.

**Comment by kristos80** [comment]
Hi @WyriHaximus thanks for the answer. Mistakenly I have started the discussion from our company's robot account.

Yes, I can confirm that I am using the `^0.2@dev` branch, but this unexpected behaviour is seen only when using the `FileInterface` with "Framework-X" and not when used alone.

I'll come back with more in the first chance.

**Reply by clue** [reply]
@kristos80 Thanks for the ping, noted your issue. I don't see how this would be caused by Framework X, so I'll wait for @WyriHaximus to come back with more details. If you can provide a minimal gist/repo to reproduce, this would be helpful in either case 👍

**Reply by kristos80** [reply]
Hi Christian @clue ,

Thanks for your response as well.

I'm quite sure that the problem is not in "Framework-X" per se, but it's the only place where I can reproduce it.

Ok, so it's getting a bit weird now.

Here's the code in a simplified form but the logic is identical:

```PHP
$filesystem = \React\Filesystem\Factory::create();
$app->get("/", function() use ($filesystem) {
	$path = __DIR__ . "/../assets/file.txt";
	await($filesystem->file($path)->putContents(microtime(TRUE) . PHP_EOL, FILE_APPEND));

	return Response::plaintext(await($filesystem->file($path)->getContents()));
});
$app->run();
```
If the file doesn't exist, then when I hit the root path and do continuous reloads, I get a response that sequentially grows as it should
```txt
1685708671.7213
1685708672.1208
...
1685708674.1959
```
Until it stops getting the content from the file that it was appending and the result stays the same.

If I restart the loop, then I get the last changes that was not able to get before and sticks there.

If I delete the file, it starts all over this pattern.

https://github.com/reactphp/reactphp/assets/13762144/67242de0-dbfa-4859-9342-a46d3cd94898

**Reply by clue** [reply]
@kristos80 Very useful details, thank you! Can you check if the file contents actually change? In other words, are we sure the reading fails (in which case this may indeed be related to the file size and the read chunk size as discussed with @WyriHaximus above) or if the appending fails?

**Reply by kristos80** [reply]
Hi @clue ,

Yes the content actually changes.

I added an `echo` statement in `STDOUT` in order to make sure is not browser caching (ok, impossible but to make sure), which had the same result with the browser output, while reading the file in OS level you could see that the new content is there.

I hope this helps :)

https://github.com/reactphp/reactphp/assets/13762144/9e08d38e-21a9-4ff3-9591-4117fad8ac1c

**Reply by clue** [reply]
@kristos80 Thank you, this appears to confirm this would likely be related to the read chunk size as discussed above. With this confirmation, let's ping @WyriHaximus, is this something you're going to look into or do we need more input? :+1:

**Reply by WyriHaximus** [reply]
@kristos80 Just for completeness sake, what are the extensions listed when you run `php -m`?

**Reply by kristos80** [reply]
@WyriHaximus sure no problem:

```text
[PHP Modules]
Core
ctype
curl
date
dom
ev
fileinfo
filter
ftp
hash
iconv
json
libxml
mbstring
mysqlnd
openssl
pcntl
pcre
PDO
pdo_mysql
pdo_sqlite
Phar
posix
readline
Reflection
session
SimpleXML
sockets
sodium
SPL
sqlite3
standard
tokenizer
xdebug
xml
xmlreader
xmlwriter
zlib

[Zend Modules]
Xdebug
```

**Reply by WyriHaximus** [reply]
@kristos80 Judging by that you are using the fallback adapter, which uses `file_get_contents`/`file_put_contents` to do reads/writes. Which makes all of this a lot more intriguing. The only difference being that `stat` call to get the file size, which I started removing in https://github.com/reactphp/filesystem/pull/115 . Can you try to patch this file by hand and see if it improves? https://github.com/reactphp/filesystem/pull/115/files#diff-4c59d68fbe70593a33b5db65e371dcff2df7f00e5fa7536a3055f4ccd78b9648 If that doesn't work, my own images coming with ext-uv and ext-eio installed, maybe that works for you: `ghcr.io/wyrihaximusnet/php:8.1-nts-alpine-slim`. If it does, we'll still fix it of course.

**Reply by odmyikona** [reply]
@WyriHaximus yes if patch the file directly with your code, it does work as expected!

**Reply by kristos80** [reply]
@WyriHaximus thanks also for the image suggestion as well. Actually I have tried to create this current image of mine by "stealing" some ideas from yours several months ago, but it seems that I didn't do a good job :)

**Reply by WyriHaximus** [reply]
> @WyriHaximus yes if patch the file directly with your code, it does work as expected!

Great! My theory is that after a few times that stat call is cached so it doesn't check anymore for a while. After a while you should start seeing the rest of the file again once the cache expires. Just a theory, tho.

> @WyriHaximus thanks also for the image suggestion as well. Actually I have tried to create this current image of mine by "stealing" some ideas from yours several months ago, but it seems that I didn't do a good job :)

Feel free to post feedback on why you didn't go with my images on my repo for it. Not sure if this thread is the best place to do that. Assuming you have any :).

**Reply by kristos80** [reply]
@WyriHaximus even after 3 hours it didn't had this sort of caching busted. I was fetching that file from within the same loop, but always the same cached result. If I restart the loop then I get the changes, but gets immediately cached as well. The only way for emptying this sort of cache is to delete/recreate the file outside of the loop, where again it would be cached after some tries to read it. Really intriguing as you said since this is the fallback mechanism.

I really am not sure why I didn't go with the exact image (probably for teaching myself on how to do that alone), but I was confident that I had ext-uv installed in the current image. This is a big lesson to learn and try not to re-invent the wheel (sorry for this off-topic)

**Reply by WyriHaximus** [reply]
> @WyriHaximus even after 3 hours it didn't had this sort of caching busted. I was fetching that file from within the same loop, but always the same cached result. If I restart the loop then I get the changes, but gets immediately cached as well. The only way for emptying this sort of cache is to delete/recreate the file outside of the loop, where again it would be cached after some tries to read it. 

@kristos80 Good to know I was wrong with that theory.

> Really intriguing as you said since this is the fallback mechanism.

Yes, but because the stat gets cache it keeps reporting the old file size. In theory you should observe the same behavior when using ext-uv or ext-eio. But working on fixing that and making sure it becomes more reliable :).

> I really am not sure why I didn't go with the exact image (probably for teaching myself on how to do that alone), but I was confident that I had ext-uv installed in the current image. This is a big lesson to learn and try not to re-invent the wheel (sorry for this off-topic)

Learning is a great reason to do things. That's how I ended up in this project ;).
